File: 0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    // map which stores copy of node
    unordered_map<Node*, Node*> copies;
    
    Node* cloneGraph(Node* node) {
        if (node == nullptr) {
            return nullptr;
        }
        // copy of first node
        Node* copy = new Node(node -> val, {});
        // stor it
        copies[node] = copy;
        queue<Node*> q;
        q.push(node);
        //bfs
        while(!q.empty()) {
            Node* curr = q.front();
            q.pop();
            for (Node* n: curr -> neighbors) {
                if (copies.find(n) == copies.end()) {
                    Node* ncopy = new Node(n -> val, {});
                    copies[n] = ncopy;
                    q.push(n);
                }
                copies[curr] -> neighbors.push_back(copies[n]);
            } 
        }
        return copy;
        

    }
};

File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node:
            return node
        node_copy = Node(node.val, [])
        copies = {node: node_copy}
        q = deque([node])
        while q:
            node = q.popleft()
            for nbr in node.neighbors:
                if nbr not in copies:
                    nbr_copy = Node(nbr.val, [])
                    copies[nbr] = nbr_copy
                    q.append(nbr)
                copies[node].neighbors.append(copies[nbr])

        return node_copy
        